clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -212,25 +214,39 @@ public RecordAppendResult append(TopicPartition tp,
      * due to metadata being unavailable
      */
-    public List<RecordBatch> abortExpiredBatches(int requestTimeout, Cluster cluster, long now) {
+    public List<RecordBatch> abortExpiredBatches(int requestTimeout,
+                                                 Cluster cluster,
+                                                 KafkaClient client,
+                                                 long now) {
         List<RecordBatch> expiredBatches = new ArrayList<RecordBatch>();
         int count = 0;
         for (Map.Entry<TopicPartition, Deque<RecordBatch>> entry : this.batches.entrySet()) {
             Deque<RecordBatch> dq = entry.getValue();
-            synchronized (dq) {
-                // iterate over the batches and expire them if they have stayed in accumulator for more than requestTimeOut
-                Iterator<RecordBatch> batchIterator = dq.iterator();
-                while (batchIterator.hasNext()) {
-                    RecordBatch batch = batchIterator.next();
-                    // check if the batch is expired
-                    if (batch.maybeExpire(requestTimeout, now, this.lingerMs)) {
-                        expiredBatches.add(batch);
-                        count++;
-                        batchIterator.remove();
-                        deallocate(batch);
-                    } else {
-                        if (!batch.inRetry()) {
-                            break;
+            // We will check if the batch should be expried if one of the following is true:
+            // 1. The leader is unknown.
+            // 2. The leader broker is disconnected.
+            Node leader = cluster.leaderFor(entry.getKey());
+            if (leader == null || client.connectionFailed(leader)) {
+                synchronized (dq) {
+                    // iterate over the batches and expire them if they have stayed in accumulator for more than requestTimeOut
+                    RecordBatch lastBatch = dq.peekLast();
+                    Iterator<RecordBatch> batchIterator = dq.iterator();
+                    while (batchIterator.hasNext()) {
+                        RecordBatch batch = batchIterator.next();
+                        // We have to close the previous batches in the deque before expiration check, otherwise those
+                        // batches may never be considered as full.
+                        if (batch != lastBatch)
+                            batch.records.close();
+                        // check if the batch is expired
+                        if (batch.maybeExpire(requestTimeout, retryBackoffMs, now, this.lingerMs)) {
+                            expiredBatches.add(batch);
+                            count++;
+                            batchIterator.remove();
+                            deallocate(batch);
+                        } else {
+                            if (!batch.inRetry()) {
```

Comment:
Good catch. I updated the patch.

There is one problem of using ```connectionFailed()```. On the first poll() if the node has not been connected yet, it will throw IllegalArgumentException. I went back and forth a little on whether adding a new ```isConnected()``` method in NetworkClient or just catching the exception. I eventually caught the exception to avoid similar methods in NetworkClient.